source/core/slang-shader-cache-index.cpp

```diff
@@ -0,0 +1,173 @@
+{
+    entryCountLimit = size;
+    shaderCacheFileSystem = fileSystem;
+    indexFilename = filename;
+
+    if (!path.getBuffer())
+    {
+        if (!shaderCacheFileSystem)
+        {
+            // Only a path was provided, so we get a mutable file system
+            // using OSFileSystem::getMutableSingleton.
+            shaderCacheFileSystem = OSFileSystem::getMutableSingleton();
+        }
+        shaderCacheFileSystem = new RelativeFileSystem(shaderCacheFileSystem, path);
+    }
+
+    // If our shader cache has an underlying file system, check if it's mutable. If so, store a pointer
+    // to the mutable version in order to save new entries later.
+    if (shaderCacheFileSystem)
+    {
+        shaderCacheFileSystem->queryInterface(ISlangMutableFileSystem::getTypeGuid(), (void**)mutableShaderCacheFileSystem.writeRef());
+    }
+
+    loadCacheIndexFromFile();
+}
+
+// Load a previous cache index saved to disk. If not found, create a new cache index
+// and save it to disk as filename.
+void ShaderCacheIndex::loadCacheIndexFromFile()
+{
+    ComPtr<ISlangBlob> indexBlob;
+    if (SLANG_FAILED(shaderCacheFileSystem->loadFile(indexFilename.getBuffer(), indexBlob.writeRef())))
+    {
+        // Cache index not found, so we'll create and save a new one.
+        if (mutableShaderCacheFileSystem)
+        {
+            mutableShaderCacheFileSystem->saveFile(indexFilename.getBuffer(), nullptr, 0);
+            return;
+        }
+        // Cache index not found and we can't save a new one due to the file system being immutable.
+        return;
+    }
+
+    String indexString;
+    File::readAllText(indexFilename, indexString);
+
+    List<UnownedStringSlice> lines;
+    StringUtil::calcLines(indexString.getUnownedSlice(), lines);
+    for (auto line : lines)
+    {
+        List<UnownedStringSlice> digests;
+        StringUtil::split(line, ' ', digests); // This will return our two hashes as two elements in digests.
+        auto dependencyDigest = stringToHash(digests[0]);
+        auto astDigest = stringToHash(digests[1]);
+
+        ShaderCacheEntry entry = { dependencyDigest, astDigest };
+        auto entryNode = entries.AddFirst(entry);
+        keyToEntry.Add(dependencyDigest, entryNode);
+    }
+}
+
+LinkedNode<ShaderCacheEntry>* ShaderCacheIndex::findEntry(const slang::Digest& key, ISlangBlob** outCompiledCode)
+{
+    LinkedNode<ShaderCacheEntry>* entryNode;
+    if (!keyToEntry.TryGetValue(key, entryNode))
+    {
+        // The key was not found in the cache, so we return nullptr.
+        *outCompiledCode = nullptr;
+        return nullptr;
+    }
+
+    // If the key is found, we need to move its corresponding entry to the front of
+    // the list and load the stored contents from disk.
+    entries.RemoveFromList(entryNode);
+    entries.AddFirst(entryNode);
+    shaderCacheFileSystem->loadFile(hashToString(key).getBuffer(), outCompiledCode);
```

Comment:
need to call `saveCacheIndexToFile()` here. We can detect if `entryNode` is already in the front of the list, and if so skip the saving step.